README.md
# Two-body cuts

![](https://img.shields.io/badge/C%2B%2B-11-green)
[![](https://img.shields.io/badge/arXiv-2106.08329-blue)](https://arxiv.org/abs/2106.08329)


Code that implements a variety of cuts for selecting two-body kinematics
at hadron colliders, including the cuts from
[arXiv:2106.08329](https://arxiv.org/abs/2106.08329).

Various examples are provided:

- [code/example1.cc](code/example1.cc): shows simple usage of the cuts,
  given the kinematics for each of two decay products. The cuts
  themselves are available through the
  [code/TwoBodyCuts.hh](code/TwoBodyCuts.hh) file, and are in the `tbc`
  namespace.

- [code/example2.cc](code/example2.cc): illustrates a Monte Carlo
  calculation of the acceptance for specific Higgs kinematics, with the
  full setup of the cuts whose acceptance shows the least dependence on
  the Higgs boson transverse momentum, compensating boost-invariant
  (CBI<sub>HR</sub>) cuts from section 5 of the paper (including an additional
  raised hardness threshold at high rapidities).

- [code/example3.cc](code/example3.cc): a similar program that evaluates
  harmonic acceptances for Z production for a variety of cuts, 
  illustrating also the use of the CBI<sub>H,DY</sub> cuts shown in section 6 of the paper.

- [code/example4.cc](code/example4.cc): a program to illustrate the use of 
  (4-phi) defiducialisation for Higgs decays with symmetric cuts.

See the start of each file for instructions on building and running the
executables.



## Dependencies

Usage of the code requires

- a C++11 compiler
- an installation of [FastJet](http://fastjet.fr)


NEWS.md
# two-body-cuts 0.2.0:
Addition of 4-phi defiducialisation functionality (Appendix D of v2 of
arXiv:2106.08329), with the new example4.cc illustrating its usage. 

# two-body-cuts 0.1.0:
First release, to coincide with appearance of arXiv article


code/example1.cc
/// example1.cc:
/// 
/// Example program to illustrate usage of cuts, given the 
/// momenta from a two-body decay
/// 
/// Build and run it with:
///
///     make example1
///     ./example1
///

// This file is part of the two-body-cuts project.
// Copyright 2021, Gavin Salam and Emma Slade (University of Oxford)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <memory>
#include "TwoBodyCuts.hh"

using namespace std;
using namespace fastjet;
// for two-body cuts
using namespace tbc;


int main(int argc, char** argv){
  // start off with some massless photon momenta
  double px1 = 40.0, py1 = 12.0, pz1 = 45.0; 
  double E1 = sqrt(px1*px1+py1*py1+pz1*pz1);
  PseudoJet gamma1(px1,py1,pz1,E1);

  double px2 = -55.0, py2 = -15.0, pz2 = -31.8; 
  double E2 = sqrt(px2*px2+py2*py2+pz2*pz2);
  PseudoJet gamma2(px2,py2,pz2,E2);

  // Create a two-body system (here generically referred to as a "Boson", 
  // though it need not actually correspond to a Boson)
  Boson boson(gamma1,gamma2);
  cout << "Boson is\n" << boson << endl;

  // Then create a range of cuts, and apply them (here we pass the cut
  // thresholds as a fraction of the boson mass and below, when we apply
  // the cut, we pass the boson divided by its mass)
  vector<CutsBase*> cuts;
  cuts.push_back(new CutsNormal  (0.25, 0.35));
  cuts.push_back(new CutsProduct (0.25, 0.35));
  cuts.push_back(new CutsCBIPt   (0.25, 0.35));
  cuts.push_back(new CutsCBIPtRap(0.25, 0.35));

  // for each of the hardness cuts, add in the info about the rapidity cuts
  for (auto & cut: cuts) {
    cut->add_upper_absrap_cut(2.37);
    cut->add_absrap_exclusion(1.37, 1.52);
  }

  for (const auto & cut: cuts) {
    cout << "Applying " << cut->description() << endl;
    // remember to divide out the boson mass if the cut thresholds passed
    // above were dimensionless
    cout << "Boson passes cuts? " << cut->pass(boson/boson.m()) << endl << endl;
  }

  // we could save this with a 
  for (auto & cut: cuts) {
    delete cut;
  } 
}


code/example2.cc
/// example2.cc:
///
/// Example program to illustrate a Monte Carlo evaluation of cut
/// efficiencies for a given choice of Higgs kinematics (i.e. Higgs pt,
/// rapidity and mass), as plotted throughout the associated paper.
///
/// Build and run it with:
///
///     make example2 ./example2
///
/// It takes a few seconds to run and gives a progress indicator as it
/// does.

// This file is part of the two-body-cuts project.
// Copyright 2021, Gavin Salam and Emma Slade (University of Oxford)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <memory>
#include <random>
#include "TwoBodyCuts.hh"

using namespace std;
using namespace fastjet;
// for two-body cuts
using namespace tbc;

int main(int argc, char** argv) {

  // first prepare the cuts that we will examine
  double yband_lo = 1.37;
  double yband_hi = 1.52;
  double ymax = 2.37;

  vector<CutsBase*> cuts;
  cuts.push_back(new CutsNormal  (0.25, 0.35));
  cuts.push_back(new CutsProduct (0.25, 0.35));

  // set up a default CBI_HR
  auto cbiHR = new CutsCBIPtRap(0.25, 0.35);
  // add in the higher ptCS cut for high rapidities
  double ytransition = (yband_lo + ymax)/2.0;
  double pthard_higher = 1.0 / 2.0 / cosh(yband_hi - ytransition);
  cbiHR->add_higher_pthard(ytransition, pthard_higher);
  cuts.push_back(cbiHR);

  // for each of the hardness cuts, add in the info about the rapidity cuts
  for (auto & cut: cuts) {
    cut->add_upper_absrap_cut(ymax);
    cut->add_absrap_exclusion(yband_lo, yband_hi);
  }
    
  // choose some Higgs kinematics at which we wish to examine the acceptace
  double ptH = 25.0;
  double yH  = 0.4;
  double mH  = 125.0;
  cout << "Boson kinematics: ptH = " << ptH 
       << ", yH = " << yH
       << ", mH = " << mH << endl;

  // choose a random number generator and a number of events.
  //
  // NB: if you want to loop over different Higgs kinematic
  //     points in order to examine the ptH dependence, it is
  //     a good idea to use the same starting seed for each
  //     ptH, so that statistical errors are correlated 
  //     across the pTH points.
  unsigned nev = 1000000;
  int seed = 5;
  mt19937 generator(seed);
  auto phi_dist      = uniform_real_distribution<double>(-pi,pi);
  auto costheta_dist = uniform_real_distribution<double>(-1.0,1.0);
  
  // do a simple MC integration to evaluate the acceptance
  // for each of the cuts (using the a common set of boson
  // decays across all the cuts)
  vector<double> npass(cuts.size(),0);
  for (unsigned iev = 0; iev < nev; iev++) {
    if (iev > 0 && iev % 100000 == 0) cout << "completed " << iev << " events out of " << nev << endl;
    // choose uniform random orientation for the Higgs decay
    double phi = phi_dist(generator);
    double theta = acos(costheta_dist(generator));
    // create boson (which by default is boosted along the x axis) and
    // evaluate whether it passes each of the cuts
    Boson boson(mH,theta,phi,ptH,yH);
    for (unsigned icut = 0; icut < cuts.size(); icut++) {
      if (cuts[icut]->pass(boson/boson.m())) npass[icut] += 1.0;
    }
  }

  // get the results
  for (unsigned icut = 0; icut < cuts.size(); icut++) {
    cout << "\nResults for \n" << cuts[icut]->description() << "\n"
         << "efficiency = " << npass[icut]/nev << "\n";
  }

}


code/example4.cc
/// example4.cc:
///
/// Example program to illustrate a Monte Carlo evaluation of cut
/// efficiencies for a given choice of Higgs kinematics (i.e. Higgs pt,
/// rapidity and mass), using events with defiducialisation.
///
/// Build and run it with:
///
///     make example4 ./example4
///
/// It takes a few seconds to run and gives a progress indicator as it
/// does.

// This file is part of the two-body-cuts project.
// Copyright 2021, Gavin Salam and Emma Slade (University of Oxford)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <memory>
#include <random>
#include "TwoBodyCuts.hh"

using namespace std;
using namespace fastjet;
// for two-body cuts
using namespace tbc;

int main(int argc, char** argv) {

  // first prepare the cuts that we will examine
  double yband_lo = 1.37;
  double yband_hi = 1.52;
  double ymax = 2.37;

  // the same transverse momentum cut applied will be applied to both
  // decay photons (or equivalently it's a cut on the soft photon pt)
  CutsDefid4Phi cuts_defid(0.25);
  // add in the info about the rapidity cuts
  cuts_defid.add_upper_absrap_cut(ymax);
  cuts_defid.add_absrap_exclusion(yband_lo, yband_hi);

  // choose some Higgs kinematics at which we wish to examine the
  // weighted acceptance
  double ptH = 25.0;
  double yH  = 0.4;
  double mH  = 125.0;
  cout << "Boson kinematics: ptH = " << ptH 
       << ", yH = " << yH
       << ", mH = " << mH << endl;

  // choose a random number generator and a number of events.
  //
  // NB: if you want to loop over different Higgs kinematic
  //     points in order to examine the ptH dependence, it is
  //     a good idea to use the same starting seed for each
  //     ptH, so that statistical errors are correlated 
  //     across the pTH points.
  unsigned nev = 1000000;
  int seed = 5;
  mt19937 generator(seed);
  auto phi_dist      = uniform_real_distribution<double>(-pi,pi);
  auto costheta_dist = uniform_real_distribution<double>(-1.0,1.0);
  
  // do a simple MC integration to evaluate the acceptance
  // for each of the cuts (using the a common set of boson
  // decays across all the cuts)
  double defiducialised_weight_no_Born = 0.0;
  double defiducialised_weight_Born = 0.0;
  for (unsigned iev = 0; iev < nev; iev++) {
    if (iev > 0 && iev % 100000 == 0) cout << "completed " << iev << " events out of " << nev << endl;
    // choose uniform random orientation for the Higgs decay
    double phi = phi_dist(generator);
    double theta = acos(costheta_dist(generator));
    // create boson (which by default is boosted along the x axis) and
    // evaluate whether it passes each of the cuts
    Boson boson(mH,theta,phi,ptH,yH);
    bool accept = cuts_defid.pass(boson/boson.m());

    if (accept) {
      // decide whether we want to defiducialise such that the weighted
      // sum over accepted events is one (false), or is equal to the 
      // Born acceptance at that rapidity (we will try both options here)
      bool normalise_to_Born = false;
      double acceptance = cuts_defid.scalar_acceptance_four_phi(boson/boson.m(), normalise_to_Born);
      defiducialised_weight_no_Born += 1.0 / acceptance;

      normalise_to_Born = true;
      acceptance = cuts_defid.scalar_acceptance_four_phi(boson/boson.m(), normalise_to_Born);
      defiducialised_weight_Born += 1.0 / acceptance;

    }
  }

  // evaluate a reference result for the expected Born acceptance
  Boson born_boson(mH,0.0,0.0,0.0,yH);
  double expected_born_acceptance = cuts_defid.scalar_acceptance_Born(born_boson/born_boson.m());

  cout << "\nResults for \n" << cuts_defid.description() << "\n"
        << "average defiducialised weight (should be equal to 1) = " << defiducialised_weight_no_Born /nev << "\n"
        << "average Born-defiducialised weight (should be equal to Born acceptance of "
        << expected_born_acceptance << ") = " << defiducialised_weight_Born /nev << "\n";

}


code/example3.cc
/// example3.cc:
///
/// Example program to illustrate a Monte Carlo evaluation of cut
/// efficiencies for a given choice of Z boson kinematics (i.e. Z pt,
/// rapidity and mass), as plotted in section 6 of the accompanying
/// paper.
///
/// Build and run it with:
///
///     make example3 ./example3
///
/// It takes a few seconds to run and gives a progress indicator as it
/// does.

// This file is part of the two-body-cuts project.
// Copyright 2021, Gavin Salam and Emma Slade (University of Oxford)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <memory>
#include <random>
#include "TwoBodyCuts.hh"

using namespace std;
using namespace fastjet;
// for two-body cuts
using namespace tbc;

// forward definition
void set_CS_weights(double theta, double phi, vector<double> & weights);

int main(int argc, char** argv) {

  double mZ  = 91.1876;

  // first prepare the cuts that we will examine
  double ymax = 2.4;

  vector<CutsBase*> cuts;
  cuts.push_back(new CutsNormal  (25.0, 25.0));
  cuts.push_back(new CutsProduct (25.0, 30.0));

  // set up a default CBI_H,DY
  auto cbiHDY = new CutsCBIPt(25.0, 30.0);
  // this sets up the cos(theta)>cbar cut (Eq. 6.7 of the paper)
  cbiHDY->set_DY_mode();
  cuts.push_back(cbiHDY);

  // for each of the hardness cuts, add in the info about the rapidity cuts
  for (auto & cut: cuts) {
    cut->add_upper_absrap_cut(ymax);
  }
    
  // choose some Higgs kinematics at which we wish to examine the acceptace
  double ptZ = 25.0;
  double yZ  = 0.4;
  cout << "Boson kinematics: ptZ = " << ptZ 
       << ", yZ = " << yZ
       << ", mZ = " << mZ << endl;

  // choose a random number generator and a number of events.
  //
  // NB: if you want to loop over different Z-boson kinematic
  //     points in order to examine the ptZ dependence, it is
  //     a good idea to use the same starting seed for each
  //     ptZ, so that statistical errors are correlated 
  //     across the pTZ points.
  unsigned nev = 10000000;
  int seed = 5;
  mt19937 generator(seed);
  auto phi_dist      = uniform_real_distribution<double>(-pi,pi);
  auto costheta_dist = uniform_real_distribution<double>(-1.0,1.0);
  
  // do a simple MC integration to evaluate the acceptance
  // for each of the cuts (using the a common set of boson
  // decays across all the cuts)
  constexpr unsigned nweights = 9;
  vector<double> weights(nweights);
  vector<vector<double>> npass(cuts.size(),vector<double>(nweights,0.0));
  for (unsigned iev = 0; iev < nev; iev++) {
    if (iev > 0 && iev % 1000000 == 0) cout << "completed " << iev << " events out of " << nev << endl;

    // choose uniform random orientation for the Z-boson decay
    double phi = phi_dist(generator);
    double theta = acos(costheta_dist(generator));

    // create boson (which by default is boosted along the x axis) and
    // determine the Collins-Soper weights
    Boson boson(mZ,theta,phi,ptZ,yZ);
    set_CS_weights(theta,phi, weights);

    // loop over cuts and weights and fill the results
    for (unsigned icut = 0; icut < cuts.size(); icut++) {
      if (cuts[icut]->pass(boson)) {
        for (unsigned iw = 0; iw < nweights; iw++) {
          npass[icut][iw] += weights[iw];
        }
      }
    }
  }

  // get the results
  for (unsigned icut = 0; icut < cuts.size(); icut++) {
    cout << "\nResults for \n" << cuts[icut]->description() << "\n";
    for (unsigned iw = 0; iw < nweights; iw++) {
      cout << "Harmonic acceptance ";
      if (iw == 8) cout << "u: ";
      else         cout << iw << ": ";
      // in the output, we include the factor of 3/4 such that
      // these accepances multiply dsigma/dq(unpol) * Ai(q)
      // (or just dsigma/dq(unpol) for i==u)
      cout << 0.75 * npass[icut][iw]/nev  << endl;
    }
  }

}

inline double pow2(double x) {return x*x;}

/// The DY cross section is given by 
///
///    dsigma/dcostheta dphi dq = 1/4pi dsigma^{unpol}/dq *
///                        3/4(weights[8] + sum_{i=0}^n weights[i] * A_i(q))
///
/// This routine takes theta and phi and fills the weights[..] vector
void set_CS_weights(double theta, double phi, vector<double> & weights){
  assert(weights.size() == 9);
  double cos_theta = cos(theta);
  double sin_theta = sin(theta);
  double cos_phi   = cos(phi);

  double sin_phi = sin(phi);
  double sin_2phi = 2*sin_phi*cos_phi;
  double cos_2phi  = 2*pow(cos_phi,2)-1;
  double sin_2theta = 2*sin_theta*cos_theta;
  weights[0] = 0.5*(1 - 3*pow2(cos_theta));
  weights[1] = sin_2theta*cos_phi;
  weights[2] = 0.5*pow2(sin_theta)*cos_2phi;
  weights[3] = sin_theta * cos_phi;
  weights[4] = cos_theta;
  weights[5] = pow2(sin_theta) * sin_2phi;
  weights[6] = sin_2theta * sin_phi;
  weights[7] = sin_theta * sin_phi;
  weights[8] = 1 + pow2(cos_theta);

}
